File: packages/java/src/main/java/fbc/FuzzCruzado.java
package fbc;

import fbc.Seguranca.CriptografiaAlvo;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HexFormat;

/**
 * Runner de fuzzing cruzado da implementação Java.
 *
 * Lê casos no formato {@code chave|iv|proposito|plaintext} (hex) e escreve uma
 * linha {@code indice|keystream|mac} por caso, para comparação byte a byte com
 * as outras linguagens do monorepo.
 */
public final class FuzzCruzado {
    private FuzzCruzado() {
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 2) {
            System.err.println("uso: FuzzCruzado <casos.txt> <saida.txt>");
            System.exit(2);
        }

        Path entrada = Path.of(args[0]);
        Path saida = Path.of(args[1]);

        CriptografiaAlvo alvo = new CriptografiaAlvo();
        HexFormat hex = HexFormat.of();

        try (BufferedReader leitor = Files.newBufferedReader(entrada, StandardCharsets.UTF_8);
                BufferedWriter escritor = Files.newBufferedWriter(saida, StandardCharsets.UTF_8)) {
            String linha;
            int indice = 0;
            while ((linha = leitor.readLine()) != null) {
                if (linha.isEmpty()) {
                    continue;
                }

                String[] partes = linha.split("\\|", -1);
                if (partes.length != 4) {
                    throw new IllegalArgumentException("Caso " + indice + " malformado: " + linha);
                }

                byte[] chave = hex.parseHex(partes[0]);
                byte[] iv = hex.parseHex(partes[1]);
                String proposito = partes[2];
                byte[] plaintext = hex.parseHex(partes[3]);

                byte[] keystream = alvo.gerarKeystreamBruto(chave, iv, proposito, plaintext.length);
                byte[] mac = alvo.checksumBruto(plaintext, chave);

                escritor.write(indice + "|" + hex.formatHex(keystream) + "|" + hex.formatHex(mac));
                escritor.newLine();
                indice++;
            }
        }
    }
}

File: packages/node/bin/fuzz_cruzado.js
'use strict';

const fs = require('fs');
const CriptografiaAlvo = require('../src/Seguranca/CriptografiaAlvo.js');

const [, , entrada, saida] = process.argv;
if (!entrada || !saida) {
  process.stderr.write('Uso: node bin/fuzz_cruzado.js <entrada> <saida>\n');
  process.exit(1);
}

let conteudo;
try {
  conteudo = fs.readFileSync(entrada, 'utf8');
} catch (erro) {
  process.stderr.write(`Não foi possível ler: ${entrada}\n`);
  process.exit(1);
}

const linhas = conteudo
  .split('\n')
  .map((linha) => linha.trim())
  .filter((linha) => linha !== '');

const alvo = new CriptografiaAlvo();
const saidaBuf = [];

linhas.forEach((linha, indice) => {
  const partes = linha.split('|');
  if (partes.length < 4) {
    process.stderr.write(`Linha malformada no índice ${indice}: ${linha}\n`);
    process.exit(1);
  }

  const [chaveHex, ivHex, proposito, plaintextHex] = partes;

  const chave = Buffer.from(chaveHex, 'hex');
  const iv = Buffer.from(ivHex, 'hex');
  const plaintext = plaintextHex === '' ? Buffer.alloc(0) : Buffer.from(plaintextHex, 'hex');

  const ks = alvo.gerarKeystreamBruto(chave, iv, proposito, plaintext.length);
  const mac = alvo.checksumBruto(plaintext, chave);

  saidaBuf.push(`${indice}|${ks.toString('hex')}|${mac.toString('hex')}`);
});

fs.writeFileSync(saida, saidaBuf.join('\n') + '\n');
